### ISSUE TITLE
Preferences Locale Problem

### ISSUE THREAD

**Issue by utnapischtim** [inveniosoftware/invenio-rdm-records#1849]
Package version (if known): current

## Describe the bug

If a user sets `Preferences Locale` to a not `en` value the worker can't create a valid json and the DOI registration on Datacite fails.

## Steps to Reproduce
1. go the the user menu page
2. set the `Preferences locale` to a not `en` value
3. create a new record
4. add a custom award
5. add a custom title to the award
6. try to publish (which works, doi creation fails)

## side bugs

- on local development if the developer stops the instance and starts again the error is gone because the instance looses the `preference locale` setting. which seems is a side bug

- the nested `_missing` should be removed so that the `json.dumps` does not fail

## Expected behavior

doi will be registered

## Problem

the `Preferences locale` setting changes the default language value which is used to store the award title. it should look like `{'title': {'en': 'fasr'}, 'number': '2345'}` but it looks like `{'title': {'de': 'fasr'}, 'number': '2345'}` . this is then a problem [here](https://github.com/inveniosoftware/invenio-rdm-records/blob/970aa7bc7f399b6bd2696aff7eb82a49f1c659c5/invenio_rdm_records/resources/serializers/datacite/schema.py#L620) the award title could not be found because there is no `en` in `title`.

to be fair, this could still create a valid object but it doesn't the nested `missing` will not be removed and therefore the `json.dumps` [here](https://github.com/inveniosoftware/datacite/blob/26f3974aef6a540a82d5aeb244c73dfa834423f4/datacite/rest_client.py#L121) will fail with `TypeError: Object of type _Missing is not JSON serializable`.

## solutions

- the solution to use `current_i18n.locale` which is used for the default locale [here](https://github.com/inveniosoftware/invenio-app-rdm/blob/36ccbaae0254fee6022130c35255c431c63e324a/invenio_app_rdm/records_ui/views/deposits.py#L330)  is not possible because the record dump to register the doi is done in the worker which is not aware of the user settings.

- remove the feature that the award title is language aware and change the serialize function accordingly

- change the serializer to `funding_ref["awardTitle"] = [award_title for _, award_title in award.get("title", {}).items()][0]`

**Comment by fenekku** [comment]
Yes approach 3) makes sense to me. To also account for the `missing` side bug, the possible absence of any title and/or lang, and split up some responsibilities in that file a bit more, the following pattern can be extended:

```python

from marshmallow import Schema, fields, missing


class FundingReferenceSchema(Schema):
    # fill with other fields ...
    awardTitle = fields.Method("get_title", data_key="title")

    def get_title(self, obj):
        """Serialize title

        :param obj: A metadata.funding dict
        :type obj: dict
        :return: serialized funding title
        :rtype: str or missing
        """
        titles_by_lang = (e for e in obj.get("title", {}).items())
        return next(titles_by_lang, missing)


class DataCite43Schema(Schema):
    fundingReferences = fields.Method("get_funding")

    def get_funding(self, obj):
        funding_list = obj["metadata"].get("funding", [])
        return FundingReferenceSchema().dump(funding_list, many=True)


obj = {
    "metadata": {
        "funding": [
            {
                "title": {
                    "en": "My title"
                }
            }
        ]
    }
}
result = DataCite43Schema().dump(obj)
print(f"result: {result}")
# result: {'fundingReferences': [{'title': 'My title'}]}

obj = {
    "metadata": {
        "funding": [
            {
                "title": {}
            }
        ]
    }
}
result = DataCite43Schema().dump(obj)
print(f"result: {result}")
# result: {'fundingReferences': [{}]}

obj = {
    "metadata": {
        "funding": [
            {
                "title": {
                    "de": "Mein titel"
                }
            }
        ]
    }
}
result = DataCite43Schema().dump(obj)
print(f"result: {result}")
# result: {'fundingReferences': [{'title': ('de', 'Mein titel')}]}
```

**Comment by ntarocco** [comment]
The `locale` part should **NOT** be changeable by the user at this time.
We implemented it to prepare the data model for the future (and avoid running data transformation/db changes in future upgrades, given that the user profile is a JSON blob) but, in the UI, it should be read-only.
I doubt there was any testing done related to locale.

The locale could/should be used for displaying purposes, but not for registering a DOI: the metadata sent should be complete and should use the exact user input, and not the current local.

**Comment by utnapischtim** [comment]
i think i found then the error.

the [serializer](https://github.com/inveniosoftware/invenio-rdm-records/blob/de1493e72c1713d57e1d83fc637021746450470f/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/serializers/FundingField.js#L103) and [deserializer](
https://github.com/inveniosoftware/invenio-rdm-records/blob/de1493e72c1713d57e1d83fc637021746450470f/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/serializers/FundingField.js#L40) in the [FundingField](https://github.com/inveniosoftware/invenio-rdm-records/blob/de1493e72c1713d57e1d83fc637021746450470f/invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/serializers/FundingField.js#L13) uses the `defaultLocale` set [here](https://github.com/inveniosoftware/invenio-app-rdm/blob/810befded43dc5ea49198bf992d6e2393c75ff1e/invenio_app_rdm/records_ui/views/deposits.py#L330)
